python/grpc_opentracing/_server.py

```diff
@@ -0,0 +1,196 @@
+"""Implementation of the service-side open-tracing interceptor."""
+
+import sys
+import logging
+import re
+
+import grpc
+from grpc_opentracing import grpcext, ActiveSpanSource, ServerRequestAttribute
+from grpc_opentracing._utilities import get_method_type, get_deadline_millis,\
+    log_or_wrap_request_or_iterator
+import opentracing
+
+
+class _OpenTracingServicerContext(grpc.ServicerContext, ActiveSpanSource):
+
+    def __init__(self, servicer_context, active_span):
+        self._servicer_context = servicer_context
+        self._active_span = active_span
+        self.code = grpc.StatusCode.OK
+        self.details = None
+
+    def is_active(self, *args, **kwargs):
+        return self._servicer_context.is_active(*args, **kwargs)
+
+    def time_remaining(self, *args, **kwargs):
+        return self._servicer_context.time_remaining(*args, **kwargs)
+
+    def cancel(self, *args, **kwargs):
+        return self._servicer_context.cancel(*args, **kwargs)
+
+    def add_callback(self, *args, **kwargs):
+        return self._servicer_context.add_callback(*args, **kwargs)
+
+    def invocation_metadata(self, *args, **kwargs):
+        return self._servicer_context.invocation_metadata(*args, **kwargs)
+
+    def peer(self, *args, **kwargs):
+        return self._servicer_context.peer(*args, **kwargs)
+
+    def send_initial_metadata(self, *args, **kwargs):
+        return self._servicer_context.send_initial_metadata(*args, **kwargs)
+
+    def set_trailing_metadata(self, *args, **kwargs):
+        return self._servicer_context.set_trailing_metadata(*args, **kwargs)
+
+    def set_code(self, code):
+        self.code = code
+        return self._servicer_context.set_code(code)
+
+    def set_details(self, details):
+        self.details = details
+        return self._servicer_context.set_details(details)
+
+    def get_active_span(self):
+        return self._active_span
+
+
+def _add_peer_tags(peer_str, tags):
+    ipv4_re = r"ipv4:(?P<address>.+):(?P<port>\d+)"
+    match = re.match(ipv4_re, peer_str)
+    if match:
+        tags['peer.ipv4'] = match.group('address')
+        tags['peer.port'] = match.group('port')
+        return
+    ipv6_re = r"ipv6:\[(?P<address>.+)\]:(?P<port>\d+)"
+    match = re.match(ipv6_re, peer_str)
+    if match:
+        tags['peer.ipv6'] = match.group('address')
+        tags['peer.port'] = match.group('port')
+        return
+    logging.warning('Unrecognized peer: \"%s\"', peer_str)
+
+
+# On the service-side, errors can be signaled either by exceptions or by calling
+# `set_code` on the `servicer_context`. This function checks for the latter and
+# updates the span accordingly.
+def _check_error_code(span, servicer_context):
+    if servicer_context.code != grpc.StatusCode.OK:
+        span.set_tag('error', True)
```

Comment:
I changed to use the constants in opentracing.ext.tags available in opentracing-1.2.2, but tags.ERROR was added for opentracing-1.2.3 which hasn't been released yet.